[Debug] Add automatic device attribute dump on composition test failures

j-ororke opened the pull request:
#### Summary

Automatically dump device attribute data when composition tests fail to improve debugging efficiency.
Previously, when conformance tests like TC_DeviceConformance or TC_DeviceBasicComposition failed, developers had to manually run IDM-12.1 to collect device attribute dumps for debugging. This PR adds automatic device data logging on test failure by:

- Moving log_structured_data() to automatically work in matter_testing.py if using --debug script-args flag.and problems are noticed during testing
- Changed inheritance order in TC_DeviceConformance and TC_DeviceBasicComposition from (MatterBaseTest, BasicCompositionTests) to (BasicCompositionTests, MatterBaseTest)
- Adding fail_current_test() override as a fallback for explicit test failures

The structured logs use clear begin/end markers making it easy to extract device data from test failure logs, eliminating the manual step of requesting IDM-12.1 dumps and significantly speeding up the debugging process.

**Note:** Currently applies to TC_DeviceConformance and TC_DeviceBasicComposition. Test failure patterns will need to be refactored if wanted to use this attribute_dump functionality moving forward as the failure pattern that is used in the TC_DeviceConformance and TC_DeviceBasicComposition tests use the self.record_error() from the BasicCompositionTests class when problems are noticed and increment a problem counter that is checked to see if the attribute dump should occur or not at the end of the test. Also requires the user to use the --debug flag in script-args.

#### Related issues

Resolves task: [680](https://github.com/project-chip/matter-test-scripts/issues/680)

#### Testing
- Verified by running this for composition tests (TC_DeviceConformance, TC_DeviceBasicComposition)
- See comment [here](https://github.com/project-chip/connectedhomeip/pull/41020#issuecomment-3575386266) for the WSL command used and a sample of the Debug output for the atribute dump.

#### Readability checklist

-   [X ] PR title is
        [descriptive](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html#title-formatting)
-   [ X] Apply the
        [_“When in Rome…”_](https://project-chip.github.io/connectedhomeip-doc/style/CODING_STYLE_GUIDE.html)
        rule (coding style)
-   [ X] PR size is short

j-ororke committed:
Add automatic device attribute dump on composition test failures:

- Move log_structured_data to module level in basic_composition support module for reusability
- Add on_fail override in BasicCompositionTests to automatically dump device data
- Ensures all composition tests automatically log device attributes on failure
- Improves debugging by providing IDM-12.1 equivalent data without manual intervention
- Add missing logging import to TC_DeviceConformance.py

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Remove duplicate on_fail() and unused import from TC_DeviceConformance

- Remove on_fail() override as it's now inherited from BasicCompositionTests
- Remove unused log_structured_data import

j-ororke committed:
Removing unused logging import

j-ororke committed:
Change inheritance order to enable automatic device attribute dump on composition test failures

- Change TC_DeviceConformance from (MatterBaseTest, DeviceConformanceTests) to (DeviceConformanceTests, MatterBaseTest)
- Change TC_DeviceBasicComposition from (MatterBaseTest, BasicCompositionTests) to (BasicCompositionTests, MatterBaseTest)
- This ensures BasicCompositionTests.teardown_class() takes precedence in Method Resolution Order
- Enables automatic device attribute dumps when composition tests detect problems
- Improves debugging efficiency by providing IDM-12.1 equivalent data without manual intervention

The inheritance order change allows the teardown_class override in BasicCompositionTests
to execute before MatterBaseTest.teardown_class, ensuring device data is dumped before
the standard problem logging occurs.

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
resolving linting errors and making sure that on test failures we emit the failure msg correctly

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Update basic_composition.py

Restoring fail_current_test() as per Cecille's suggestion, thank you!

j-ororke committed:
Update basic_composition.py

Updated to using module-level logger instance instead of using logging directly

j-ororke committed:
Merge branch 'master' into create_global_attr_log_function

restyled-commits committed:
Restyled by autopep8

cecille commented on src/python_testing/matter_testing_infrastructure/matter/testing/basic_composition.py:
```diff
@@ -278,3 +295,27 @@ def build_spec_xmls(self):
         self.xml_device_types, problems = build_xml_device_types(dm)
         self.problems.extend(problems)
+
+    def teardown_class(self):
```
Does this work? The basic information / composition tests do that double-inheritance thing. I was thinking this would happen on the MatterBaseTest teardown.

Though we can actually do proper inheritance now, so if you want to change this, that could make sense too.

j-ororke replied:
Hey Cecille,

You're totally right - I was overcomplicating this with the inheritance gymnastics! 😅

I've now moved the device attribute dumping functionality from BasicCompositionTests.teardown_class() into MatterBaseTest.teardown_class() where I feel it belongs. 
This way:

- No more multiple inheritance headaches
- All Matter tests get the debugging goodness automatically
- Much cleaner code that follows proper inheritance patterns
- Everything still works exactly as before

Thanks for steering me in the right direction! The device dumps will now help debug failures across the entire test suite, not just the composition tests.

Much appreciated!

j-ororke committed:
Move device attribute dumping from BasicCompositionTests to MatterBaseTest for universal test debugging

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Merge branch 'master' into create_global_attr_log_function

cecille commented on src/python_testing/TC_DeviceBasicComposition.py:
```diff
@@ -604,12 +604,14 @@ class RequiredMandatoryAttribute:
                     if suffix > attribute_standard_range_max and suffix < global_range_min:
                         self.record_error(self.get_test_name(), location=location,
-                                          problem=f"Manufacturer attribute in undefined range {manufacturer_value} in cluster {cluster_id}",
+                                          problem=f"Manufacturer attribute in undefined range {
```
We're about to update the restyler so this doesn't happen - there was a bug in the restyler that is causing this weirdness. When that lands, would you mind reverting these changes? I think this might break python prior to 3.12.

cecille commented on src/python_testing/matter_testing_infrastructure/matter/testing/matter_testing.py:
```diff
@@ -222,8 +222,54 @@ def teardown_class(self):
+                    _, txt_str = self.dump_wildcard(None)
+                    # Only dump the text format - it's more readable for debugging
+                    self._log_structured_data('==== FAILURE_DUMP_txt: ', txt_str)
+                else:
```
I don't know if it's worthwhile logging these - it might just confuse people. Ditto with the exception.

cecille commented on src/python_testing/matter_testing_infrastructure/matter/testing/matter_testing.py:
```diff
@@ -222,8 +222,54 @@ def teardown_class(self):
             LOGGER.info("###########################################################")
+
+            # Attempt to dump device attribute data for debugging when problems are found
+            self._dump_device_attributes_on_failure()
```
Can you please move this to above the problem dump? That will make the actual problems appear next to the failure message rather than forcing people to scroll up to above the device dump.

j-ororke replied:
Updated, moved the problem dump to above the problem logging.

j-ororke committed:
Merge branch 'master' into create_global_attr_log_function

j-ororke committed:
Fix Python 3.11 compatibility and improve test failure User Experience:

- Fix f-strings in TC_DeviceBasicComposition.py for Python < 3.12
- Move device dump before problem logging for better failure visibility
- Remove noisy log messages from device dump on failure
- Use specific exceptions instead of catching all exceptions so that it isnt blanketed.

cecille commented on src/python_testing/matter_testing_infrastructure/matter/testing/matter_testing.py:
```diff
@@ -224,6 +227,42 @@ def teardown_class(self):
+            # Don't let data access or serialization errors interfere with the original test failure
+            pass
+
+    def _log_structured_data(self, start_tag: str, dump_string: str):
```
This seems like a duplicate of the function in basic_composition.py

j-ororke committed:
removed duplicate log_structured_data function from basic_composition module, now inheriting it directly from matter_testing in test_TC_IDM_12_1 test

j-ororke committed:
removed duplicate log_structured_data function from basic_composition module, now inheriting it directly from matter_testing in test_TC_IDM_12_1 test in TC_DeviceBasicCompositon test module

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Resolving linting error

cecille reviewed (commented):
So thinking this through a bit more - I'm wondering if we should flag this functionality so we're not dropping device details into test logs. I suppose in some ways the tests already expose information. But maybe with a debug bool-arg or something? WDYT? Maybe something to bring to the TH team?

j-ororke committed:
Apply suggestions from code review from Cecille

Co-authored-by: C Freeman <cecille@google.com>

j-ororke committed:
Merge branch 'master' into create_global_attr_log_function

j-ororke committed:
Adding --debug flag to do attribute dump during conformance and composition tests only when added in script-args, also added super().teardown_class() to device conformance and composition test modules

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Resolve mypy issue

j-ororke committed:
Adding CI test runs with debug flag

cecille commented on src/python_testing/TC_DeviceBasicComposition.py:
```diff
@@ -167,6 +167,18 @@
 #       --trace-to perfetto:${TRACE_TEST_PERFETTO}.perfetto
 #     factory-reset: true
 #     quiet: true
+#   run17:
```
I don't think this will actually do anything - all the tests pass in the CI so you won't hit your code.

cecille replied:
ah, it's fine - I think the underlying functions are already tested.

cecille commented on src/python_testing/TC_DeviceBasicComposition.py:
```diff
@@ -284,13 +297,17 @@ def check_no_duplicates(obj: Any) -> None:
         await self.setup_class_helper()
         self.build_spec_xmls()
 
+    @async_test_body
```
Does adding this do anything? Does the super not already get called if you don't override?

j-ororke committed:
Removed teardown_class override as inheritance from MatterBaseTest appears to be working fine now
